### PR TITLE
unify type/references creation

### PR DESCRIPTION
**Pull request by anz-gordonj7**
This PR pulls out almost all the type parsing into a seperate child parser as a further step to simplifying listener_impl.go. The reason for this is:
1) Ensure consistency around usage of types in the parser, anywhere which expects a type should allow any type of type.
2) Centralise all type creation so that TypeRef types can be removed from the parsed sysl file (so eval code can use the real sysl.Type instead of guessing the type via strings.

@anz-bank/sysl-developers

### CODE REVIEW

File: pkg/parse/tests/foreign_import_swagger.sysl.golden.textpb [hunk 1]
```diff
@@ -131,6 +131,16 @@ apps: <
           query_param: <
             name: "depth"
             type: <
+              type_ref: <
```
Comment (sahejsingh) [inline pkg/parse/tests/foreign_import_swagger.sysl.golden.textpb]: These changes ( golden.textpb changes without corresponding .sysl changes) are regressions and not compatible with legacy python.
Need to be looked at more closely.

Comment (anz-gordonj7) [reply pkg/parse/tests/foreign_import_swagger.sysl.golden.textpb]: compatibility is not a priority, anz-bank/sysl-legacy is where that belongs.

File: pkg/parse/resolvetyperef_test.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		Ref: &sysl.Scope{
+			Appname: &sysl.AppName{
+				Part: t.appname,
+			},
+			Path: t.path,
+		},
+	}
+}
+
+func (t typereftestdata) Run(tt *testing.T, mod *sysl.Module) {
```
Comment (joshcarp) [inline pkg/parse/resolvetyperef_test.go]: Comment

File: pkg/parse/typeparser.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		Type: &sysl.Type_TypeRef{
+			TypeRef: &sysl.ScopedRef{
+				Context: syslContext,
+				Ref:     ref,
+			},
+		},
+	}
+}
+
+func (t *typeParser) EnterName_str(ctx *parser.Name_strContext) {
```
Comment (joshcarp) [inline pkg/parse/typeparser.go]: Comment

File: pkg/parse/typeparser.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	*t.namestrTarget = append(*t.namestrTarget, ctx.GetText())
+}
+
+func (t *typeParser) EnterUser_defined_type(ctx *parser.User_defined_typeContext) {
```
Comment (joshcarp) [inline pkg/parse/typeparser.go]: Comment

File: pkg/parse/typeparser.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	t.namestrTarget = &t.typePath
+}
+
+func (t *typeParser) EnterReference(ctx *parser.ReferenceContext) {
```
Comment (joshcarp) [inline pkg/parse/typeparser.go]: Comment

File: pkg/parse/typeparser.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	t.namestrTarget = &t.appName
+}
+
+func (t *typeParser) ExitApp_name(ctx *parser.App_nameContext) {
```
Comment (joshcarp) [inline pkg/parse/typeparser.go]: Comment

**Comment by anzdaddy** [comment]
@joshcarp Let me have a go at changing your mind. In Go, comments are predominantly used for documenting exported symbols. Everywhere else, they should be used as sparingly as possible, mostly relying on the clarity of the code itself to document the intent.

In this instance, the symbols are exported only by necessity because they implement the methods of the `SyslParserListener` callback interface and is not intended for consumption by a developer.

**Comment by anz-gordonj7** [comment]
@joshcarp are you asking for comments on those functions because "they are exported" or because you the code doesnt make it clear what it is doing?

**Pull request closed without merging**